src/relay/tcprelay/crypto_io.rs

```diff
@@ -255,14 +207,56 @@ where
     S: AsyncWrite + Unpin,
 {
     fn poll_write(self: Pin<&mut Self>, ctx: &mut Context<'_>, buf: &[u8]) -> Poll<io::Result<usize>> {
-        self.priv_poll_write(ctx, buf)
+        let this = self.get_mut();
+        match this.cipher.category() {
+            CipherCategory::None => {
+                Pin::new(&mut this.stream).poll_write(ctx, buf)
+            },
+            CipherCategory::Stream => {
+                // NOTE: API 经过良好设计的话，此处不需要 Copy 数据。
+                let mut payload = [0u8; 1 << 14]; // 16K
+                let buf_len = buf.len();
+                payload[..buf_len].copy_from_slice(buf);
+
+                this.cipher.encrypt_packet(&mut payload[..buf_len]);
+
+                ready!(Pin::new(&mut this.stream).poll_write(ctx, &payload[..buf_len]))?;
+
+                Poll::Ready(Ok(buf_len))
+            },
+            CipherCategory::Aead => {
+                /// AEAD packet payload must be smaller than 0x3FFF
+                const MAX_PACKET_SIZE: usize = 0x3FFF;
+                assert!(buf.len() <= MAX_PACKET_SIZE);
+
+                let tag_len = this.cipher.tag_len();
+                let hlen = 2 + tag_len;
+
+                let mut payload = [0u8; 1 << 14]; // 16K
+                let buf_len = buf.len() as u16;
+                payload[0..2].copy_from_slice(&(buf_len.to_be_bytes()));
+                
+                this.cipher.encrypt_packet(&mut payload[..hlen]);
+
+                let offset = hlen + buf.len();
+                payload[hlen..offset].copy_from_slice(buf);
+
+                let offset = offset + tag_len;
+                this.cipher.encrypt_packet(&mut payload[hlen..offset]);
+
+                // Send
+                ready!(Pin::new(&mut this.stream).poll_write(ctx, &payload[..offset]))?;
```

Comment:
Same as above. `poll_write` is the same as `write()` syscall, it will return if system's snd_buf is full.